wolfcrypt/src/port/nxp/dcp_port.c

```diff
@@ -0,0 +1,537 @@
+    dcp_unlock();
+
+    return ret;
+}
+
+int wc_InitSha256(wc_Sha256* sha256)
+{
+    return wc_InitSha256_ex(sha256, NULL, INVALID_DEVID);
+}
+
+void DCPSha256Free(wc_Sha256* sha256)
+{
+    if (sha256)
+        dcp_free(sha256->handle.channel);
+}
+
+int wc_Sha256Update(wc_Sha256* sha256, const byte* data, word32 len)
+{
+    int ret;
+    if (sha256 == NULL || (data == NULL && len != 0)) {
+        return BAD_FUNC_ARG;
+    }
+    dcp_lock();
+    ret = DCP_HASH_Update(DCP, &sha256->ctx, data, len);
+    if (ret != kStatus_Success)
+        ret = WC_HW_E;
+    dcp_unlock();
+    return ret;
+}
+
+int wc_Sha256GetHash(wc_Sha256* sha256, byte* hash)
+{
+    int ret;
+    size_t outlen = WC_SHA256_DIGEST_SIZE;
+    dcp_hash_ctx_t saved_ctx;
+    if (sha256 == NULL || hash == NULL)
+        return BAD_FUNC_ARG;
+    dcp_lock();
+    XMEMCPY(&saved_ctx, &sha256->ctx, sizeof(dcp_hash_ctx_t));
+    XMEMSET(hash, 0, WC_SHA256_DIGEST_SIZE);
+    ret = DCP_HASH_Finish(DCP, &sha256->ctx, hash, &outlen);
+    if ((ret != kStatus_Success) || (outlen != SHA256_DIGEST_SIZE))
+        ret = WC_HW_E;
+    else
+        XMEMCPY(&sha256->ctx, &saved_ctx, sizeof(dcp_hash_ctx_t));
+    dcp_unlock();
+    return 0;
+}
+
+int wc_Sha256Final(wc_Sha256* sha256, byte* hash)
+{
+    int ret;
+    size_t outlen = WC_SHA256_DIGEST_SIZE;
+    dcp_lock();
+    ret = DCP_HASH_Finish(DCP, &sha256->ctx, hash, &outlen);
+    if ((ret != kStatus_Success) || (outlen != SHA256_DIGEST_SIZE))
+        ret = WC_HW_E;
+    else {
+        ret = DCP_HASH_Init(DCP, &sha256->handle, &sha256->ctx, kDCP_Sha256);
+        if (ret < 0)
+            ret = WC_HW_E;
+    }
+    dcp_unlock();
+    return ret;
+}
+
+#if defined(WOLFSSL_HASH_FLAGS) || defined(WOLF_CRYPTO_CB)
+int wc_Sha256SetFlags(wc_Sha256* sha256, word32 flags)
+{
+    if (sha256) {
+        sha256->flags = flags;
+    }
+    return 0;
+}
+int wc_Sha256GetFlags(wc_Sha256* sha256, word32* flags)
+{
+    if (sha256 && flags) {
+        *flags = sha256->flags;
+    }
+    return 0;
+}
+#endif /* WOLFSSL_HASH_FLAGS || WOLF_CRYPTO_CB */
+
+int wc_Sha256Copy(wc_Sha256* src, wc_Sha256* dst)
+{
+    if (src == NULL || dst == NULL)
+        return BAD_FUNC_ARG;
+    dcp_lock();
+    XMEMCPY(&dst->ctx, &src->ctx, sizeof(dcp_hash_ctx_t));
+    dcp_unlock();
+    return 0;
+}
+#endif /* !NO_SHA256 */
+
+
+#ifndef NO_SHA
+
+int wc_InitSha_ex(wc_Sha* sha, void* heap, int devId)
+{
+    int ret;
+    int ch;
+    int keyslot;
+    if (sha == NULL)
+        return BAD_FUNC_ARG;
+    ch = dcp_get_channel();
+    if (ch == 0)
+        return WC_PENDING_E;
+    keyslot = dcp_key_slot(ch);
+    dcp_lock();
+    (void)devId;
+    XMEMSET(sha, 0, sizeof(wc_Sha));
+    sha->handle.channel    = ch;
+    sha->handle.keySlot    = keyslot;
+    sha->handle.swapConfig = kDCP_NoSwap;
+    ret = DCP_HASH_Init(DCP, &sha->handle, &sha->ctx, kDCP_Sha1);
+    if (ret != kStatus_Success)
+        ret = WC_HW_E;
+    dcp_unlock();
+    return ret;
+}
```

Comment:
```suggestion
int wc_InitSha_ex(wc_Sha* sha, void* heap, int devId)
{
    int ret;
    int ch;
    if (sha == NULL)
        return BAD_FUNC_ARG;
    ch = dcp_get_channel();
    if (ch == 0)
        return WC_PENDING_E;
    dcp_lock();
    (void)devId;
    XMEMSET(sha, 0, sizeof(wc_Sha));
    sha->handle.channel    = ch;
    sha->handle.keySlot    = kDCP_KeySlot0;
    sha->handle.swapConfig = kDCP_NoSwap;
    ret = DCP_HASH_Init(DCP, &sha->handle, &sha->ctx, kDCP_Sha1);
    if (ret != kStatus_Success)
        ret = WC_HW_E;
    dcp_unlock();
    return ret;
}
```
Key slot is not used in hash, just set to default value